src/devices/cpu/mipsx/mipsxdasm.cpp

```diff
@@ -0,0 +1,588 @@
+			}
+			break;
+
+		}
+		case 6:
+		{
+			// stf - Store Floating Point
+			// stf Offset[rSrc1], rDest
+			util::stream_format(stream, "stf %08x[%s], %s", imm17, get_regname(get_src1(opcode)), get_regname(get_src2_dest(opcode)));
+			break;
+		}
+		case 7: // movtoc
+		{
+			// this is just a suggested form
+			u8 c2 =   BIT(opcode, 0, 4);
+			u8 c1 =   BIT(opcode, 4, 4);
+			u8 func = BIT(opcode, 8, 6);
+			u8 op =   BIT(opcode, 14, 3);
+			u8 dest = get_src2_dest(opcode);
+			u8 src =  get_src1(opcode);
+
+			if (src == 0)
+			{
+				util::stream_format(stream, "movtoc %s, (%02x, %02x, %02x, %02x)", get_regname(dest), op, func, c1, c2);
+			}
+			else
+			{
+				//util::stream_format(stream, "illegal movtoc form");
+				// this form appears to be used
+				util::stream_format(stream, "movtoc %s, (%02x, %02x, %02x, %02x) (src1 == %s)", get_regname(dest), op, func, c1, c2, get_regname(src));
+			}
+			break;
+
+		}
+		default:
+		{
+			util::stream_format(stream, "Unhandled TY2 (%08x)", opcode);
+			break;
+		}
+		}
+		break;
+	}
+
+	case 3:
+	{
+		u8 op = get_op(opcode);
+
+		switch (op)
+		{
+		case 0:
+		{
+			// jspci - Jump Indexed and Store PC
+			// jspci rSrc1,#Immed,Rdest
+			int imm17 = get_imm17(opcode);
+			imm17 = util::sext(imm17, 17) * 2;
+			util::stream_format(stream, "jspci %s, #%02x, %s", get_regname(get_src1(opcode)), imm17, get_regname(get_src2_dest(opcode)));
+			break;
+		}
+
+		case 1:
+		{
+			// Halt and Spontaneously Combust
+			if ((opcode & 0x07ffffff) == 0x07c00000)
+			{
+				util::stream_format(stream, "hsc");
+			}
+			else
+			{
+				util::stream_format(stream, "illegal hsc form");
+			}
+			break;
+		}
+
+		case 2:
+		{
+			// movtos - Move to Special Register
+			// movtos rSrc1, SpecialReg
+
+			// psw  001
+			// md   010
+			// pcm1 100
+			u16 comp = get_compute_compfunc(opcode);
+			if ((comp & 0xffe) == 0x000)
+			{
+				u8 dest = get_src2_dest(opcode);
+
+				if (dest == 0x00)
+				{
+					u8 src = get_src1(opcode);
+					util::stream_format(stream, "movtos %s, %01x", get_regname(src), comp & 0x7);
+				}
+				else
+				{
+					util::stream_format(stream, "illegal movtos form");
+				}
+			}
+			else
+			{
+				util::stream_format(stream, "illegal movtos form");
+			}
+			break;
+
+		}
+
+		case 3:
+		{
+			// movfrs - Move from Special Register
+			// movfrs SpecialReg, rDest
+
+			// psw  001
+			// md   010
+			// pcm4 100
+			u16 comp = get_compute_compfunc(opcode);
+			if ((comp & 0xffe) == 0x000)
+			{
+				u8 src = get_src1(opcode);
+				if (src == 0x00)
+				{
+					u8 dest = get_src2_dest(opcode);
+					util::stream_format(stream, "movfrs %01x, %s", comp & 0x7, get_regname(dest));
+				}
+				else
+				{
+					util::stream_format(stream, "illegal movfrs form");
+				}
+			}
+			else
+			{
+				util::stream_format(stream, "illegal movfrs form");
+			}
+			break;
+
+		}
+
+		case 4:
+		{
+			int imm17 = get_imm17(opcode);
+			imm17 = util::sext(imm17 & 0x1ffff, 17);
```

Comment:
Same as previous comment - `util::sext()` should be sufficient. What's the mask for?

Reply:
while redundant, I prefer the form, as it makes it easier to remember which bit the sign bit is, as I often forget.  I'll change it anyway if that is preferred, but I was going for readability.